File: frontend/src/types/feed.ts
export interface RssFeed {
  id: string;
  name: string;
  url: string;
  feed_type: 'rss' | 'web_scrape' | 'twitter';
  css_selector: string | null;
  competitor_id: string | null;
  competitor_name: string | null;
  is_active: boolean;
  last_checked_at: string | null;
  last_successful_at: string | null;
  error_count: number;
  last_error: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  // Twitter-specific fields (populated when feed_type === 'twitter')
  x_username?: string;
  x_user_id?: string;
  backfill_completed?: boolean;
  include_retweets?: boolean;
  include_replies?: boolean;
}

File: backend/models/feed.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import String, Boolean, Integer, Text, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.database import Base


class RSSFeed(Base):
    __tablename__ = "rss_feeds"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String, nullable=False)
    url: Mapped[str] = mapped_column(String, unique=True, nullable=False)
    competitor_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("competitors.id"), nullable=True
    )
    feed_type: Mapped[str] = mapped_column(String, default="rss", server_default="rss", nullable=False)
    css_selector: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    last_checked_at: Mapped[datetime | None] = mapped_column(nullable=True)
    last_successful_at: Mapped[datetime | None] = mapped_column(nullable=True)
    error_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    last_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(default=func.now(), onupdate=func.now(), nullable=False)

    competitor = relationship("Competitor", back_populates="feeds")
    creator = relationship("User")
    items = relationship("FeedItem", back_populates="feed")

